synapse/storage/databases/main/event_federation.py

```diff
@@ -1530,6 +1536,78 @@ def _record_event_failed_pull_attempt_upsert_txn(
 
         txn.execute(sql, (room_id, event_id, 1, self._clock.time_msec(), cause))
 
+    @trace
+    async def filter_event_ids_with_pull_attempt_backoff(
+        self,
+        room_id: str,
+        event_ids: Collection[str],
+    ) -> List[str]:
+        """
+        Filter down the events to ones that we've failed to pull before recently. Uses
+        exponential backoff.
+
+        Args:
+            room_id: The room that the events belong to
+            event_ids: A list of events to filter down
+
+        Returns:
+            List of event_ids that should not be attempted to be pulled
+        """
+
+        def _filter_event_ids_with_pull_attempt_backoff_txn(
+            txn: LoggingTransaction,
+        ) -> List[str]:
+            where_event_ids_match_clause, values = make_in_list_sql_clause(
+                txn.database_engine, "event_id", event_ids
+            )
+
+            if isinstance(self.database_engine, PostgresEngine):
+                least_function = "least"
+            elif isinstance(self.database_engine, Sqlite3Engine):
+                least_function = "min"
+            else:
+                raise RuntimeError("Unknown database engine")
+
+            sql = f"""
+                SELECT event_id FROM event_failed_pull_attempts
+                WHERE
+                    room_id = ?
+                    AND {where_event_ids_match_clause}
+                    /**
+                    * Exponential back-off (up to the upper bound) so we don't try to
+                    * pull the same event over and over. ex. 2hr, 4hr, 8hr, 16hr, etc.
+                    *
+                    * We use `1 << n` as a power of 2 equivalent for compatibility
+                    * with older SQLites. The left shift equivalent only works with
+                    * powers of 2 because left shift is a binary operation (base-2).
+                    * Otherwise, we would use `power(2, n)` or the power operator, `2^n`.
+                    */
+                    AND ? /* current_time */ < last_attempt_ts + (
```

Comment:
Also, TBH I'm wondering whether for this it would be clearer to just fetch `event_id, last_attempt_ts, num_attempts` and do the checks in python?

Reply:
> Also, TBH I'm wondering whether for this it would be clearer to just fetch `event_id, last_attempt_ts, num_attempts` and do the checks in python?

I think this makes a lot of sense here since we have to get a response out for each `event_ids` anyway 👍


<br>

---



> It occurs to me that these inline SQL comments get sent over the wire to postgres, which is probably not totally ideal. Might be worth moving the big ones out?

In terms of other places to worry about this: we already [strip out the newlines](https://github.com/matrix-org/synapse/blob/b6baa46db078c3ef9e6c5751bccb8d2e1c5c5402/synapse/storage/database.py#L401-L403) which is why I use `/* */` instead of inline `--`. Is there a way we can strip the comments out so we don't lose the nice readability? The performance of find/replacing this every time doesn't sound great but our `_make_sql_one_line` doesn't look particularly optimized by eye either 🤔. I'm not sure why we started doing that (introduced in https://github.com/matrix-org/synapse/pull/1939), maybe to get rid of all of the extra indentation spaces in multi-line strings.

Reply:
Created https://github.com/matrix-org/synapse/issues/14192 to continue this comments in SQL conversation ⏩

> Note that we don't actually send the one-line SQL to the database, we use it for logging.

huh, I swore this was a problem before. Good to know!